### ISSUE TITLE
Sipi hangs and reports bad file descriptors

### ISSUE THREAD

**Issue by benjamingeer** [dhlab-basel/Sipi#84]
This is a load test of Sipi, using the nginx web server to simulate the Knora API server. Nginx just serves a static JSON file.

Steps to reproduce on OS X:

1. `brew install nginx`
2. Pull the Sipi `tests/scala_test` branch.
3. Start nginx (in the Sipi directory):
```
mkdir -p it/src/test/resources/nginx/logs
nginx -p it/src/test/resources/nginx/ -c nginx.conf
```
4. Check that nginx is simulating Knora by opening [http://127.0.0.1:3434/v1/files/Leaves.jpg](http://127.0.0.1:3434/v1/files/Leaves.jpg) in a browser.
6. Verify that nginx handles 500 concurrent requests very fast, with no errors:
```
ab -n 10000 -c 500 http://127.0.0.1:3434/v1/files/Leaves.jpg
```
7. Start Sipi so it uses nginx:
```
local/bin/sipi -config config/sipi.fake-knora-test-config.lua
```
8. Check that Sipi is working with nginx by opening [http://127.0.0.1:1024/knora/Leaves.jpg/full/full/0/default.jpg](http://127.0.0.1:1024/knora/Leaves.jpg/full/full/0/default.jpg) in a browser.
9. Make 100 concurrent requests to Sipi:
```
ab -n 10000 -c 100 http://127.0.0.1:1024/knora/Leaves.jpg/full/full/0/default.jpg
```

Results:

After a few thousand requests, Sipi hangs, and cannot be stopped with Control-C.

Then `sipi.log` contains errors like this:

```
[2016-12-12 19-43-27] [SIPI:ERROR] server.http() failed: Error #1279: Could not connect socket! Bad file descriptor
[2016-12-12 19-43-27] [SIPI:ERROR] GET: /knora/Leaves.jpg/full/full/0/default.jpg failed: Unknown error! Unauthorized access!
```

If I try with only 10 concurrent requests, I get this:

```
$ ab -n 5000 -c 10 http://localhost:1024/knora/Leaves.jpg/full/full/0/default.jpg
This is ApacheBench, Version 2.3 <$Revision: 1706008 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient)
Completed 500 requests
Completed 1000 requests
apr_socket_recv: Connection reset by peer (54)
Total of 1109 requests completed
```

Then `sipi.log` contains:

```
[2016-12-12 20-09-30] [SIPI:ERROR] server.http() failed: Error #1279: Could not connect socket! Bad file descriptor
[2016-12-12 20-09-30] [SIPI:ERROR] GET: /knora/Leaves.jpg/full/full/0/default.jpg failed: Unknown error! Unauthorized access!
[2016-12-12 20-09-30] [SIPI:ERROR] Commpipe_write close error: Bad file descriptor
```

To stop nginx:

```
nginx -p it/src/test/resources/nginx/ -c nginx.conf -s stop
```

**Comment by benjamingeer** [comment]
Doing this doesn't help:

```
sudo sysctl kern.ipc.somaxconn=12000
sudo sysctl net.inet.tcp.msl=1000
sudo sysctl net.inet.ip.portrange.first=32768
```

**Comment by benjamingeer** [comment]
I made another test with no Knora, just a Lua script that always returns `allow` (so it doesn't use Knora or nginx or anything).

With 10 or 100 concurrent requests everything seems fine:

```
ab -n 10000 -c 100 http://127.0.0.1:1024/knora/incunabula_0000003846.jpg/full/full/0/default.jpg
```

So I think the problem is in Sipi's HTTP client code, not in its HTTP server code.

However, the number of concurrent requests that Sipi will accept is still limited by the backlog parameter of `listen`, which is set in `Server.cpp` to `SOMAXCONN`, which is set to 128 in `/usr/include/sys/socket.h`. I think a higher value would be more reasonable. Although 128 could be enough for normal usage, there might be bursts of higher load. In a load test, the backlog must be high enough to accept the initial burst of requests.

**Comment by benjamingeer** [comment]
How about using [libcurl](https://curl.haxx.se/libcurl/) as an HTTP client library?

**Comment by benjamingeer** [comment]
I added `libcurl` to the build on the `tests/scala_test` branch, but haven't yet tried using it in `LuaServer.cpp`.

**Comment by benjamingeer** [comment]
I replaced the custom HTTP client implementation with `libcurl`, and now I can do 10 or 100 concurrent requests with no errors.

**Comment by benjamingeer** [comment]
Clearly my error handling isn't quite right. If the client connection fails, I get:

```
SHTTPS-ERROR at [/Users/benjamingeer/git/Sipi/shttps/LuaServer.cpp: #2461] 
```

**Comment by benjamingeer** [comment]
Fixed in #90.
